ls1intum/Artemis
# `Programming exercises`: Fix missing build plan during import on Jenkins setups

### Checklist
#### General


#### Server

### Motivation and Context


Closes #7056 

### Description

In order to easily access the exercises' build plans more easily in the `JenkinsBuildPlanService`, the `copyBuildPlan` method in the `ContinuousIntegrationService` interface is changed to pass the complete programming exercise objects instead of only the project keys. The keys can still be easily retrieved from the exercises by calling the `getProjectKey` method. In addition, a new method is added to the `BuildPlanRepository`. That one is used to actually copy the new build plan from source to target for the Jenkins setups.

Fixed import process:
- Connect the build plan in the Artemis database additionally to the new exercise.
- In Jenkins: make sure the URL from which the build plan is downloaded (Artemis REST-API) is updated to link to the new programming exercise ID and secret token.
  - Otherwise the build plan of the original exercise would still be used.
  - This is done directly as part of the copy build plan process, i.e., Artemis fetches the old build plan, updates the necessary URL, and copies is back into the correct directory in Jenkins.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- test server with GitLab/Jenkins setup

1. Log in to Artemis
2. Navigate to Course Management
3. Click `Exercises`
4. Click `Import Programming Exercise` and make sure `From Instance` is selected.
5. Click the import button next to the existing exercise.
6. Fill in the required fields and save the exercise.
7. Click on the `Edit Build Plan` button.
8. Make sure that the build plan is present and is the same as the one from the source exercise.

#### Exam Mode Testing



### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage





| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExerciseImportBasicService.java | 100% | ✅                           |
| ProgrammingExerciseImportService.java | 98% | ✅                           |
| JenkinsBuildPlanService | 94% | ✅                           |
